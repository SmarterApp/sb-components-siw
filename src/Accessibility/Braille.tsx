import * as React from "react";

function getCookie(name: string): string | undefined {
  const nameLenPlus = name.length + 1;

  return (
    document.cookie
      .split(";")
      .map(c => c.trim())
      .filter(cookie => {
        return cookie.substring(0, nameLenPlus) === `${name}=`;
      })
      .map(cookie => {
        return decodeURIComponent(cookie.substring(nameLenPlus));
      })[0] || undefined
  );
}

export interface BrailleLinkProps {
  currentSelectionCode: string;
  brailleItemCodes?: string[];
  braillePassageCodes?: string[];
  bankKey: number;
  itemKey: number;
}

export interface BrailleLinkState {
  displaySpinner: boolean;
}
export class BrailleLink extends React.Component<
  BrailleLinkProps,
  BrailleLinkState
> {
  constructor(props: BrailleLinkProps) {
    super(props);
    this.state = {
      displaySpinner: false
    };
  }

  buildUrl(bankKey: number, itemKey: number): string {
    let url = "";
    const { brailleItemCodes } = this.props;

    if (
      brailleItemCodes &&
      brailleItemCodes.indexOf(this.props.currentSelectionCode) > -1
    ) {
      const brailleLoc = brailleItemCodes.indexOf(
        this.props.currentSelectionCode
      );
      const brailleType = brailleItemCodes[brailleLoc];

      url = `/Item/Braille?bankKey=${bankKey}&itemKey=${itemKey}&brailleCode=${brailleType}`;
    }

    return url;
  }

  enableSpinner(): void {
    this.setState({
      displaySpinner: true
    });
  }

  disableSpinner(): void {
    this.setState({
      displaySpinner: false
    });
  }

  checkDownloadCookie(count: number) {
    const innerCount = count + 1;
    if (getCookie("brailleDLstarted") === "1") {
      this.disableSpinner();
      document.cookie =
        "brailleDLstarted" + "=;expires=Thu, 01 Jan 1970 00:00:00 GMT;path=/;";

      return;
    }
    if (innerCount > 20) {
      document.cookie =
        "brailleDLstarted" + "=;expires=Thu, 01 Jan 1970 00:00:00 GMT;path=/;";
      this.disableSpinner();

      return;
    }
    const dlCheck = this.checkDownloadCookie;
    setTimeout(dlCheck.bind(this), 1000, innerCount);

    return;
  }

  watchForDlStart(): void {
    this.enableSpinner();
    this.checkDownloadCookie(0);
  }

  renderLoading() {
    let content;
    if (this.state.displaySpinner) {
      content = (
        <span className="glyphicon glyphicon-refresh glyphicon-pad rotating" />
      );
    }

    return content;
  }

  render() {
    const brailleUrl = this.buildUrl(this.props.bankKey, this.props.itemKey);
    let content;
    if (brailleUrl !== "") {
      content = (
        <a
          className={"item-nav-btn btn btn-default btn-sm about-pt-btn"}
          aria-live="polite"
          aria-relevant="additions"
          href={brailleUrl}
          download
          onClick={() => this.watchForDlStart()}
        >
          <span className="fa fa-download" />
          Download Braille Embossing File(s)
          {this.renderLoading()}
        </a>
      );
    }

    return content;
  }
}
