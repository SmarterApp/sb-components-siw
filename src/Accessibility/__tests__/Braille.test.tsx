import * as React from "react";
import { shallow } from "enzyme";
import { BrailleLink, BrailleLinkState } from "../Braille";

describe("Braille", () => {
  const wrapper = shallow(
    <BrailleLink
      currentSelectionCode="test"
      brailleItemCodes={["1", "2", "test"]}
      braillePassageCodes={["1", "2"]}
      bankKey={111}
      itemKey={222}
    />
  );

  it("renders correctly", () => {
    wrapper.findWhere(node => node.type() === BrailleLink);
    expect(wrapper).toMatchSnapshot();
  });

  it("no render", () => {
    const noRenderWrapper = shallow(
      <BrailleLink
        currentSelectionCode="nonMatchingCode"
        brailleItemCodes={["1", "2"]}
        braillePassageCodes={["1", "2"]}
        bankKey={111}
        itemKey={222}
      />
    );

    expect(noRenderWrapper).toMatchSnapshot();
  });

  it("refresh state", () => {
    const newState: BrailleLinkState = { displaySpinner: true };
    const newWrapper = wrapper.setState(newState);

    expect(newWrapper).toMatchSnapshot();
    expect(newWrapper.find(".glyphicon-refresh")).toMatchSnapshot();
  });
});
