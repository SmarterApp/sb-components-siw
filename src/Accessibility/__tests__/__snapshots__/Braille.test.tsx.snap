// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Braille no render 1`] = `""`;

exports[`Braille refresh state 1`] = `
<a
  aria-live="polite"
  aria-relevant="additions"
  className="item-nav-btn btn btn-default btn-sm about-pt-btn"
  download={true}
  href="/Item/Braille?bankKey=111&itemKey=222&brailleCode=test"
  onClick={[Function]}
>
  <span
    className="fa fa-download"
  />
  Download Braille Embossing File(s)
  <span
    className="glyphicon glyphicon-refresh glyphicon-pad rotating"
  />
</a>
`;

exports[`Braille refresh state 2`] = `
<span
  className="glyphicon glyphicon-refresh glyphicon-pad rotating"
/>
`;

exports[`Braille renders correctly 1`] = `
<a
  aria-live="polite"
  aria-relevant="additions"
  className="item-nav-btn btn btn-default btn-sm about-pt-btn"
  download={true}
  href="/Item/Braille?bankKey=111&itemKey=222&brailleCode=test"
  onClick={[Function]}
>
  <span
    className="fa fa-download"
  />
  Download Braille Embossing File(s)
</a>
`;
